lib/day_3/binary_diagnostic.ex

```diff
@@ -0,0 +1,116 @@
+defmodule AdventOfCode2021.BinaryDiagnostic do
+  @moduledoc """
+  Advent of Code 2021
+  Day 3: Binary Diagnostic
+  """
+
+  @spec power_consumption :: number
+  def power_consumption do
+    {_, puzzle} = get_puzzle()
+    data = %{gamma_rate: "", epsilon_rate: ""}
+    # Get map with index and count values to compare
+    puzzle
+    |> puzzle_symbol_range()
+    |> Enum.map(&bit_count(puzzle, &1))
+    # Compose gamma and epsilon strings
+    |> compose_characteristics(data)
+    # Put values to list and calculate multiply result
+    |> calculate_result()
+  end
+
+  @spec life_support_rating :: integer
+  def life_support_rating do
+    data = %{puzzle: elem(get_puzzle(), 1)}
+
+    with oxygen <- get_rating_value(data.puzzle, {data, :common}),
+         scrubber <- get_rating_value(data.puzzle, {%{puzzle: elem(get_puzzle(), 1)}, :less}) do
+      oxygen * scrubber
+    end
+  end
+
+  defp get_rating_value(puzzle, {data, type}) when type in [:common, :less] do
+    puzzle
+    |> puzzle_symbol_range()
+    # Find dominant or least bit and apply criteria filter
+    |> define_rating_bit({data, type})
+    # Get value from map, parse to decimal
+    |> extract_result()
+  end
+
+  defp define_rating_bit(puzzle_symbol_range, {data, type}) do
+    Enum.reduce_while(puzzle_symbol_range, data, fn index, acc ->
+      acc =
+        acc.puzzle
+        |> bit_count(index)
+        |> rating_bit(type)
+        |> apply_criteria_filter({acc, index})
+
+      if length(acc.puzzle) > 1, do: {:cont, acc}, else: {:halt, acc}
+    end)
+  end
+
+  defp apply_criteria_filter(bit, {acc, index}) do
+    %{acc | puzzle: Enum.filter(acc.puzzle, fn row -> String.at(row, index) == bit end)}
+  end
+
+  defp extract_result(data) do
```

Comment:
Too abstract name, ideally should reflect what is going here

Reply:
renamed as `extract_characteristic_rating`